File: acwing/算法提高课/第一章动态规划/1064.小国王/1064.小国王.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>
#include <vector>

using namespace std;

const int N = 12, M = 1 << 10, K = 110;

typedef long long LL;

LL f[N][K][M];
int cnt[M];
vector<int> state;
vector<int> he[M];
int n, m;

bool check(int state)
{
    if ((state >> 1) & state)
    {
        return false;
    }
    return true;
}

int count(int state)
{
    int res = 0;
    for (int i = 0; i < n; i++)
        res += (state >> i) & 1;
    return res;
}

int main()
{
    scanf("%d%d", &n, &m);
    for (int i = 0; i < (1 << n); i++)
    {
        if (check(i))
        {

            state.push_back(i);
            cnt[i] = count(i);
        }
    }
    for (int i = 0; i < state.size(); i++)
    {
        for (int j = 0; j < state.size(); j++)
        {
            int a = state[i], b = state[j];
            if ((a & b) == 0 && check(a | b))
                he[i].push_back(j);
        }
    }
    f[0][0][0] = 1;
    for (int i = 1; i <= n + 1; i++)
    {
        for (int j = 0; j < m + 1; j++)
        {
            for (int a = 0; a < state.size(); a++)
            {
                for (int b : he[a])
                {
                    int c = cnt[state[a]];
                    if (j >= c)
                    {
                        f[i][j][state[a]] += f[i - 1][j - c][state[b]];
                    }
                }
            }
        }
    }
    printf("%lld", f[n + 1][m][0]);
    return 0;
}

File: acwing/算法提高课/第一章动态规划/1064.小国王/1064.小国王.py
N = 12
K = N * N
M = 1 << 10

f = [[[0 for _ in range(M)] for _ in range(K)]for _ in range(N)]
state = []
he = [[] for _ in range(M)]
cnt = [0] * M
n, k = map(int, input().split())


def check(state):
    if state >> 1 & state:
        return False
    return True


def count(state):
    res = 0
    for i in range(n):
        res += state >> i & 1
    return res


for i in range(1 << n):
    if check(i):
        state.append(i)
        cnt[i] = count(i)

for i in range(len(state)):
    for j in range(len(state)):
        a = state[i]
        b = state[j]
        if a & b == 0 and check(a | b):
            he[i].append(j)

f[0][0][0] = 1
for i in range(n + 2):
    for j in range(k + 1):
        for a in range(len(state)):
            for b in range(len(he[a])):
                if j >= cnt[state[a]]:
                    f[i][j][state[a]] += f[i - 1][j -
                                                  cnt[state[a]]][state[he[a][b]]]

print(f[n + 1][k][0])
